[MU3] fix #303422: lyrics color is preserved when importing mxml + small bugfix and cleanup for isEven

SKefalidis opened the pull request:
Relates to: https://musescore.org/en/node/303422

I removed the `_even` variable which does not do anything other than creating a bug where there are lyrics that are neither odd nor even (as far as `Lyrics::layout` is concerned). Even `Layout::isEven` uses `_no` instead of `_even`.

This further breaks the behavior of the program when importing lyrics (since now no lyrics have the correct color when imported), but I would think that it is a step in the right direction :^).

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [x] I created the test (mtest, vtest, script test) to verify the changes I made

SKefalidis commented:
As far as the issue is concerned, I see 2 possible solutions. One is removing the call to `styleChanged` from the Layout method. As far as I can tell, no other class calls `styleChanged` from its Layout method, and this removal fixes the problem. On the other hand, even with this fix, when the user changes the style of the score, the color of the lyrics gets reset. So the other (probably better solution) is to track down that bug which might solve this too.

Jojo-Schmitz commented:
Strange enough `Lyrics::isEven()` uses `_no % 1` whereas `Lyrics::propertyDefault()` uses `_no & 1`. Should be the same thing, but the latter might be faster

Jojo-Schmitz commented:
Strange enough `Lyrics::isEven()` uses `_no % 1` whereas `Lyrics::propertyDefault()` and `Lyrics::layout()`  use `_no & 1`. Should be the same thing, but the latter might be faster, so you might as well change that too

SKefalidis commented:
The travis failure is to be expected.

Jojo-Schmitz commented:
I believe that Travis CI mtest failure is exactly what this change is supposed to fix? See https://travis-ci.org/github/musescore/MuseScore/jobs/682247601#L4340-L4362

Jojo-Schmitz commented:
So this PR does related to an earlier PR or your's, #5910, right?

SKefalidis commented:
Yes, it does relate to that PR. 
I looked around a bit more and I believe that removing the call to `styleChanged` is the way to go (I'll test it more and I will upload it). This should fix the related issue.
Also `_no % 1` is also a bug if I am not mistaken, so you are right, I'll change that.

Jojo-Schmitz commented:
Outch I didn't even notice that it is `_no % 1` but read if for `_no % 2`

mattmcclinch commented:
The purpose of the `_even` flag is to determine whether or not the parity of the verse number has changed since the last time `Lyrics::layout()` was called. `initTid()` should only be called when the lyric changes from even to odd or vice versa. It certainly should not be called every time the lyric is laid out. The `_even` flag makes sure that it is only called when necessary. I do not see what removing this flag is supposed to fix.

SKefalidis commented:
@mattmcclinch that is very interesting and useful (and your valuable information helps explain the regression described above). What it fixed is the inconsistent behavior between odd and even lyrics. It looks like the problem lies somewhere else.

Jojo-Schmitz commented:
How about `setNo(int n)` not just setting `_no`, but also `_even`?

SKefalidis commented:
My guess is that the problem is that all lyrics are initialized using the first constructor, therefore the yare all considered odd. Then layout runs and the even ones become even and they lose their properties because of styleChanged.

mattmcclinch commented on importexport/musicxml/importmxmlpass2.cpp:
```diff
@@ -737,6 +737,7 @@ static void addLyric(MxmlLogger* logger, const QXmlStreamReader* const xmlreader
             }
       else {
             l->setNo(lyricNo);
+            l->setEven(lyricNo % 2);
```
Really? The % operator?

SKefalidis replied:
I would prefer to do the following `l->setEven(l->isEven())` but doesn't that look even more confusing? Either way, hopefully, @Jojo-Schmitz idea of setting `_even` inside `setNo` will work flawlessly. That will alleviate the need for `setEven`.

Jojo-Schmitz replied:
Why not use  `l->setEven(isEven(lyricsNo))`?

mattmcclinch replied:
I believe that setting `_even` inside `setNo()` is a very bad idea. Unless you mean you are going to compare it against its previous value and call `initTid()` and set `styleDidChange` to `true` if the values differed. Probably best to just let this all happen in one place.

SKefalidis replied:
@mattmcclinch yes, that's what I was thinking to do. `styleDidChange` is also out of place (since it's usually, if not always) not in *::layout. Frankly, I don't love either option. I prefer setters than only set, but at the same time having to call a second function when setting the number of the lyric is asking for trouble.

mattmcclinch replied:
Sorry, what are the two options that you don't love?

SKefalidis replied:
The first one is setting `_even` in `setNo`. The second one is having to call 2 functions `setNo` and `setEven` together.

mattmcclinch replied:
Then let's not do either. Leave it the way it is and just change `% 2` to `& 1`.

SKefalidis commented:
Finally, a solution that seems pretty safe and decent (thank you very much @mattmcclinch). I added a comment that explains the use of _even.

@Jojo-Schmitz I like your idea very much, but at first I would like to test if that breaks something else (like the capella import).

SKefalidis commented:
Added comments, changed `% 2` -> `& 1`.

Jojo-Schmitz commented on libmscore/lyrics.cpp:
```diff
@@ -260,12 +260,14 @@ void Lyrics::layout()
 
       bool styleDidChange = false;
-      if ((_no & 1) && !_even) {
+      if (isEven() && !_even) {
```
Hmm, now this does look confusing. One would need to look up what `isEven()` does to find that it does not work on `_even`, but on '_no' 
I know I siggested to use `isEven()`, but I change my mind in the light of how this looks now ;-)

SKefalidis replied:
I feel like the real problem is `_even`. Should I change it to `_wasEven`?

Jojo-Schmitz replied:
But then `void setEven(bool b)                            { _even = b; }` seems strange

Jojo-Schmitz commented on libmscore/lyrics.cpp:
```diff
@@ -260,12 +260,14 @@ void Lyrics::layout()
             styleDidChange    = true;
             }
-      if (!(_no & 1) && _even) {
+      else if (!isEven() && _even){
```
same here

Jojo-Schmitz commented on libmscore/lyrics.cpp:
```diff
@@ -556,7 +558,7 @@ QVariant Lyrics::propertyDefault(Pid id) const
       switch (id) {
             case Pid::SUB_STYLE:
-                  return int((_no & 1) ? Tid::LYRICS_EVEN : Tid::LYRICS_ODD);
+                  return int(isEven() ? Tid::LYRICS_EVEN : Tid::LYRICS_ODD);
```
here though it is OK

SKefalidis commented:
Should I also modify testLyricsColor.xml to cover this case (of multiple colored lines of lyrics)?
(@Jojo-Schmitz )

Jojo-Schmitz commented:
> Should I also modify testLyricsColor.xml to cover this case (of multiple colored lines of lyrics)?

Yes, I think so

SKefalidis commented:
Modified the testLyricsColor.xml file to cover this case. Renamed `_wasEven` to _evenLyricStyle`. Added a couple of assertions to protect against a future misuse.

Jojo-Schmitz commented:
Travis CI's mtest is not happe with the new test file, see https://travis-ci.org/github/musescore/MuseScore/jobs/684349891#L4348-L4574

Jojo-Schmitz commented:
Rebase needed. Maybe one for 3.x and one for master?

SKefalidis commented:
> Rebase needed. Maybe one for 3.x and one for master?

On it

SKefalidis commented:
Update: there is a much bigger problem. The color of all lyrics, and probably every style change, is reset when opening the file (xml). commit 0a11f687c2b40c29fe39c10c3d1f3eae594086ec and commit 583c71cd50c1d90bb6ce0339919312008bbd0c5f  seem to be the ones that caused this issue.

SKefalidis commented:
Update: Since I haven't been following the latest developments for 3.6 I don't really know how to tackle the problem. From what I've gathered when opening a score in MuseScore there is a style change. That change causes the loss of style info (like color). There seems to be special handling for .mscx files but not for xml files which are being reset to the defaults. The best I could come up with is to add a preference to disable the automatic style migration when an xml file is being loaded but I imagine that this is not something that is wanted. Anyway it would be a separate PR. This PR accomplishes its goal, but there is a bigger problem that somebody who is more involved in the handling of XML files should tackle. I have marked the relevant commits in a previous comment.

MarcSabatella commented:
I can't say for sure if this is related, but it is true that after a score is loaded, there is a call to styleChanged() that wasn't there before.  This *should* be harmless, but if there were any styled properties that were read and *not* explicitly set to unstyled, this would cause the setting to be lost on the call to styleChanged().  Even before this change, however, It would normally *also* have been lost on save/reload, and possibly on copy/paste.  Also, certainly, if you actually *did* change a style setting.  So if you see any of those symptoms, that's the cause, and the fix is to be sure to set the property flags to unstyled when setting them,.

Jojo-Schmitz commented on libmscore/lyrics.h:
```diff
@@ -82,8 +82,9 @@ class Lyrics final : public TextBase {
+      void setEvenLyricStyle(bool b);
       int no() const                                  { return _no; }
-      bool isEven() const                             { return _no % 1; }
+      bool isEven() const                             { return _no & 1; }
```
See also #7628 and #7627

SKefalidis replied:
1600 PRs later and at least a part of this has been merged

Jojo-Schmitz replied:
And then even only that single line :-/

Jojo-Schmitz commented:
Seems this PR doesn't solve the issue, not with 3.6 anymore, there colored lyrics are not imported at all

SKefalidis commented:
In a previous comment, I've marked the commits that introduced the problem.

Jojo-Schmitz commented:
Ah, sorry, I missed those

Jojo-Schmitz commented:
So clearly a 3.6 regression. @vpereverzev ?

Jojo-Schmitz commented on libmscore/lyrics.cpp:
```diff
@@ -155,6 +155,20 @@ bool Lyrics::readProperties(XmlReader& e)
+
+void Lyrics::setEvenLyricStyle(bool b)
+      {
+      if (b == true)
```
Here I'd rather use just `if (b)`

Jojo-Schmitz replied:
Or even just  a 1-lined `Q_ASSERT(b ? isEven() : !isEven());`

Jojo-Schmitz commented:
See #8033 and #8034, I guess with those this one here can get closed
